File: src/main/java/at/yawk/math/data/Inequation.java
package at.yawk.math.data;

import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Value;

/**
 * @author yawkat
 */
@Value
public class Inequation implements Condition {
    private final Expression lhs;
    private final Equality equality;
    private final Expression rhs;

    @Override
    public String toString(int radix) {
        return lhs.toString(radix) + equality.getSymbol() + rhs.toString(radix);
    }

    @RequiredArgsConstructor
    @Getter
    public enum Equality {
        EQUAL("="),
        NOT_EQUAL("!="),
        LESS_THAN("<"),
        GREATER_THAN(">"),
        LESS_THAN_OR_EQUAL("<="),
        GREATER_THAN_OR_EQUAL(">=");

        private final String symbol;
    }
}

File: src/main/kotlin/at/yawk/math/data/Inequation.kt
package at.yawk.math.data

import at.yawk.math.EqualsHelper

/**
 * @author yawkat
 */
class Inequation(left: Expression, val equality: Equality, right: Expression) : BinaryExpression(left, right), Condition {
    override fun toString(lhs: String, rhs: String): String {
        return "$lhs ${equality.symbol} $rhs"
    }

    override fun equals(other: Any?): Boolean {
        return EqualsHelper.equals<Inequation>(other, { it.left == left && it.equality == equality && it.right == right })
    }

    override fun hashCode(): Int {
        return EqualsHelper.hashCode(left, equality, right)
    }

    protected override fun withChildren(left: Expression, right: Expression): BinaryExpression {
        return Inequation(left, equality, right)
    }
}
